index.js

```diff
@@ -430,35 +432,48 @@ module.exports = class Hypercore extends EventEmitter {
   }
 
   replicate (isInitiator, opts = {}) {
+    const { firewall, onfirewalled } = this
+    opts = { firewall, onfirewalled, ...opts }
     // Only limitation here is that ondiscoverykey doesn't work atm when passing a muxer directly,
     // because it doesn't really make a lot of sense.
-    if (Protomux.isProtomux(isInitiator)) return this._attachToMuxer(isInitiator, opts)
+    if (Protomux.isProtomux(isInitiator)) return this._attachToMuxer(isInitiator, { session: true, ...opts })
 
     const protocolStream = Hypercore.createProtocolStream(isInitiator, opts)
     const noiseStream = protocolStream.noiseStream
     const protocol = noiseStream.userData
-    const useSession = !!opts.session
 
-    this._attachToMuxer(protocol, useSession)
+    this._attachToMuxer(protocol, opts)
 
     return protocolStream
   }
 
-  _attachToMuxer (mux, useSession) {
+  _attachToMuxer (mux, opts = {}) {
     if (this.opened) {
-      this._attachToMuxerOpened(mux, useSession)
+      this._attachToMuxerOpened(mux, opts)
     } else {
-      this.opening.then(this._attachToMuxerOpened.bind(this, mux, useSession), mux.destroy.bind(mux))
+      this.opening.then(this._attachToMuxerOpened.bind(this, mux, opts), mux.destroy.bind(mux))
     }
 
     return mux
   }
 
-  _attachToMuxerOpened (mux, useSession) {
+  _attachToMuxerOpened (mux, opts = {}) {
     // If the user wants to, we can make this replication run in a session
     // that way the core wont close "under them" during replication
-    const session = useSession ? this.session() : null
-    this.replicator.attachTo(mux, session)
+    const useSession = !!opts.session
+
+    const maybeAttach = (deny) => {
+      if (deny) return opts.onfirewalled(mux, this)
+      const session = useSession ? this.session() : null
+      this.replicator.attachTo(mux, session)
+    }
+
+    if (!opts.firewall) return maybeAttach(false)
+
+    mux.stream.noiseStream.opened.then(() => {
+      const { remotePublicKey } = mux.stream.noiseStream
+      opts.firewall(remotePublicKey, this).then(maybeAttach, mux.destroy.bind(mux))
```

Comment:
If we are unable to successfully return a firewall verdict, it feels right to panic on the mux and destroy.